Refactor HTMLTreeBuilder to use direct stack operations instead of processFakeEndTag

kbolisetty opened the pull request:
#### 837aac0123734947aba1bf1877d2cb218fbca542
<pre>
Refactor HTMLTreeBuilder to use direct stack operations instead of processFakeEndTag
<a href="https://bugs.webkit.org/show_bug.cgi?id=301541">https://bugs.webkit.org/show_bug.cgi?id=301541</a>
<a href="https://rdar.apple.com/157176196">rdar://157176196</a>

Reviewed by NOBODY (OOPS!).

Replace processFakeEndTag() with direct element stack manipulation in
closeTheCell() and processEndTag() for optgroup. This eliminates token
creation overhead and state machine traversal, making the code more
explicit and efficient.

Tests: fast/parser/optgroup-end-tag-processing.html
       fast/parser/table-cell-implicit-close.html

* LayoutTests/fast/parser/optgroup-end-tag-processing-expected.txt: Added.
* LayoutTests/fast/parser/optgroup-end-tag-processing.html: Added.
* LayoutTests/fast/parser/table-cell-implicit-close-expected.txt: Added.
* LayoutTests/fast/parser/table-cell-implicit-close.html: Added.
* PerformanceTests/Parser/optgroup-implicit-close.html: Added.
* PerformanceTests/Parser/table-cell-implicit-close.html: Added.
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::closeTheCell):
(WebCore::HTMLTreeBuilder::processEndTag):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/837aac0123734947aba1bf1877d2cb218fbca542

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128921 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1175 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39752 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136300 "Built successfully") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130792 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1117 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1053 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98148 "Passed tests") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131868 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/848 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115482 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78775 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/779 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33597 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79580 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109216 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34091 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138767 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/981 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/952 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106683 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1039 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111820 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106496 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/805 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30345 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/53441 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1054 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64370 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/892 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/948 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/987 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/fast/parser/optgroup-end-tag-processing.html:
```diff
@@ -0,0 +1,137 @@
+    log('Test optgroup end tag processing in HTMLTreeBuilder');
+    log('');
+
+    // Test 1: Optgroup with unclosed option
```
I don't think we need these comments give the log right beneath it.

rniwa commented on LayoutTests/fast/parser/optgroup-end-tag-processing.html:
```diff
@@ -0,0 +1,137 @@
+    var optgroups1 = select1.querySelectorAll('optgroup');
+    var options1 = select1.querySelectorAll('option');
+
+    if (optgroups1.length === 2 && options1.length === 3) {
```
No curly braces around a single line statement: https://webkit.org/code-style-guidelines/#braces-one-line

rniwa commented on LayoutTests/fast/parser/optgroup-end-tag-processing.html:
```diff
@@ -0,0 +1,137 @@
+    // Verify proper nesting
+    var group1Options = optgroups1[0].querySelectorAll('option');
+    var group2Options = optgroups1[1].querySelectorAll('option');
+    if (group1Options.length === 2 && group2Options.length === 1) {
```
Instead of manually checking the condition and printing out PASS/FAIL log, we should use `shouldBe` function:
```js
shouldBe(`optgroups1[0].querySelectorAll('option').length`, `2`);
shouldBe(`optgroups1[1].querySelectorAll('option').length`, `1`);
```

rniwa commented on LayoutTests/fast/parser/optgroup-end-tag-processing.html:
```diff
@@ -0,0 +1,137 @@
+    var options1 = select1.querySelectorAll('option');
+
+    if (optgroups1.length === 2 && options1.length === 3) {
+        log('PASS: Created 2 optgroups and 3 options');
```
This doesn't tell us why that behavior is correct. Ideally, we should make the code inside `shouldBe` self explanatory so that we don't have to write English sentences to explain why this is correct.

rniwa commented on LayoutTests/fast/parser/optgroup-end-tag-processing.html:
```diff
@@ -0,0 +1,137 @@
+if (window.testRunner)
+    testRunner.dumpAsText();
+
+function log(msg) {
```
Use `../../resources/js-test.js`. See other tests which uses that script.

Commit:
Refactor HTMLTreeBuilder to use direct stack operations instead of processFakeEndTag
https://bugs.webkit.org/show_bug.cgi?id=301541
rdar://157176196

Reviewed by NOBODY (OOPS!).

Replace processFakeEndTag() with direct element stack manipulation in
closeTheCell() and processEndTag() for optgroup. This eliminates token
creation overhead and state machine traversal, making the code more
explicit and efficient.

Tests: fast/parser/optgroup-end-tag-processing.html
       fast/parser/table-cell-implicit-close.html

* LayoutTests/fast/parser/optgroup-end-tag-processing-expected.txt: Added.
* LayoutTests/fast/parser/optgroup-end-tag-processing.html: Added.
* LayoutTests/fast/parser/table-cell-implicit-close-expected.txt: Added.
* LayoutTests/fast/parser/table-cell-implicit-close.html: Added.
* PerformanceTests/Parser/optgroup-implicit-close.html: Added.
* PerformanceTests/Parser/table-cell-implicit-close.html: Added.
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::closeTheCell):
(WebCore::HTMLTreeBuilder::processEndTag):
